Objects/regions.c

```diff
@@ -670,8 +720,277 @@ PyObject* _Py_MakeImmutable(PyObject* obj)
+        PyObject *item = stack_pop(info.pending);
+        PyTypeObject *type = Py_TYPE(item);
+        regionmetadata *item_region = _Py_CAST(regionmetadata *, Py_REGION(item));
+
+        if (Py_REGION(item) == region) {
+            // -1 for the refernce we just followed
+            region_data->lrc -= 1;
+            continue;
+        }
+
+        if (IS_DEFAULT_REGION(Py_REGION(item))) {
+            // Add reference to the object,
+            // minus one for the reference we just followed
+            region_data->lrc += item->ob_refcnt - 1;
+            Py_SET_REGION(item, flagged_region);
+            item_region = region_data;
+
+            if (PyFunction_Check(item)) {
+                regionerror err = {.src = _PyObject_CAST(region_data->bridge),
+                    .tgt = item, .id = ERR_WIP_FUNCTIONS };
+                if ((handle_error)(&err, handle_error_data)) {
+                    continue;
+                } else {
+                    goto fail;
+                }
+            } else {
+                traverseproc traverse = type->tp_traverse;
+                if (traverse != NULL) {
+                    if (traverse(item, (visitproc)_add_to_region_visit, &info)) {
+                        stack_free(info.pending);
+                        return NULL;
+                    }
+                }
+            }
+
+            PyObject* type_op = _PyObject_CAST(Py_TYPE(item));
+            if (Py_REGION(type_op) != region && !_Py_IsImmutable(type_op)) {
+                if (stack_push(info.pending, type_op))
+                {
+                    stack_free(info.pending);
+                    return PyErr_NoMemory();
+                }
+            }
+
+            continue;
+        }
+
+        if (_Py_IsImmutable(item)) {
+            continue;
+        }
+
+        // C wrappers can propergate through the entire system and draw
+        // in a lot of unwanted objects. Since c wrappers don't have mutable
+        // data, we just make it immutable and have the immutability impl
+        // handle it. We then have an edge from our region to an immutable
+        // object which is again valid.
+        if(is_c_wrapper(item)) {
+            _Py_MakeImmutable(item);
+            continue;
+        }
+
+        if (!is_bridge_object(item)) {
+            // FIXME: This uses the bridge object for now, since we
+            // currently don't store the reference source. We need
+            // to figure out where we can get the source from since the
+            // stack only holds one object. Or maybe we don't even need
+            // the source object?
+            regionerror err = {.src = _PyObject_CAST(region_data->bridge),
+                               .tgt = item, .id = ERR_INTER_REGION_REF };
+            if ((handle_error)(&err, handle_error_data)) {
+                continue;
+            } else {
+                goto fail;
+            }
+        }
+
+        if (item_region->parent != NULL) {
```

Comment:
If I understand the code correctly, this checks that the object is a bridge object that is not the root of a region tree, and the error on 956 is that we are breaking external uniqueness. First I had trouble understanding the difference between the error on 946 and 956 since they are both illegal inter-region references. If I am correct, then this error is only possible if the source object is in the local region -- and the problem is that we are trying to promote a borrowed reference to an owned reference. I suggest the error label should be either related to external uniqueness or borrowing.

Reply:
Maybe it would make sense to do this check around line 907 to avoid pushing an object to the stack if it is a borrowed reference to a "parented" region object since we don't want to traverse the object regardless. Not sure.

Reply:
Nice catch, I'll try to rewrite it to move more things into the `_add_to_region_visit` which should make several things easier and fix this part.